### DISCUSSION TITLE
Helper tool for Application/Singleton Scoped Dependencies

### DISCUSSION THREAD

**Discussion by cal-pratt** [fastapi/fastapi#9215] [Show and tell]
Hi all, I wanted to share a project I started working on which can be integrated into FastAPI applications.
I wouldn't consider this project production ready yet, but was hoping to get some feedback from the FastAPI community.

A lot of this work was started because I wanted a way to easily create singleton or application scoped dependencies in my FastAPI application and not define them as module level globals. This was to allow easily mocking components during testing, and avoid having module loads create stateful items. Examples of where I wanted this were database connection-pools and [Kafka producers](https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/asyncio_example.py).

I also wanted to reuse my FastAPI codebase in my Kafka consumer applications while still being able to use dependency injection. This is not easily done outside the context of a webrequest, which my Kafka consumer applications do not have access to.

So I started a library that I'm calling [Flex DI](https://flexdi.readthedocs.io/en/latest/index.html) to perform dependency injection in a similar way to FastAPI, in that dependencies can be defined as easy to understand provider functions, while supporting a concept of scopes (application vs request scoped).

Gluing it back together with FastAPI was actually quite painless. [This example](https://flexdi.readthedocs.io/en/latest/fastapi.html) shows how to create a SQLAlchemy engine object as an application scoped dependency without having it defined as a global in the module. Everything that requires request state can continue using the FastAPI DI system, but more generic objects can be offloaded to Flex DI, keeping them more easily used in separate non web-based applications.

```py
from typing import AsyncIterator

from fastapi import FastAPI
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncConnection, AsyncEngine, create_async_engine

from flexdi.fastapi import FastAPIGraph, FlexDepends

app = FastAPI()
graph = FastAPIGraph(app)

@graph.bind(scope="application", eager=True)
async def provide_engine() -> AsyncIterator[AsyncEngine]:
    engine = create_async_engine("sqlite+aiosqlite:///mydb.db")
    try:
        yield engine
    finally:
        await engine.dispose()

@graph.bind
async def provide_connection(engine: AsyncEngine) -> AsyncIterator[AsyncConnection]:
    async with engine.begin() as conn:
        yield conn

@app.get("/")
async def get(conn: AsyncConnection = FlexDepends(AsyncConnection)) -> list[str]:
    statement = text("SELECT name FROM sqlite_master;")
    return [table_name for [table_name] in await conn.execute(statement)]
```

The integration with FastAPI is fairly basic at this point, but seems to act as I want it to. Here's [the implementation](https://github.com/cal-pratt/flexdi/blob/e6bc9456a6b37e85817e635fc4c6921e2248d78d/src/flexdi/fastapi/__init__.py). I tie the startup of the dependency injection system to the startup and shutdown events of FastAPI, and within that startup create an application scope for dependency injection.

Each time a new request comes in, a request scope is created and attached to the request object, which allows resolving dependencies defined as `FlexDepends`. The `FlexDepends` is a simple factory function for creating `fastapi.Depends` instances, configured to fetch objects from Flex DI.

I would love to hear what people think, if this could be improved upon in terms of usability, or any other general concerns.

P.S. Thank you @tiangolo for creating such an amazing framework.
It has improved my productivity creating webservices in Python dramatically!!

**Comment by mike-optifino** [comment]
Hi! I have similar problem when start using Celery with FastAPI. It's very difficult to reuse the same concept/solution for dependency injection in celery task and provide consistency in all code base. I wrote my own implementation of how resolve and inject dependencies for task but will be happy if FastAPI provide ability to reuse it and will not need to create my own bike. My example of code for celery task:

```
from fastapi import Depends

class Downloader: 
    def download(self, task_id):
        ...
    
IDownloader = Annotated[Downloader, Depends(Downloader)]

class DownloadIllustrationTaskArgs(TaskArgs):
    task_id: UUID

@async_task(
    name='download_illustration',
    autoretry_for=(
        SeleniumTimeoutException,
    ),
    retry_kwargs={'max_retries': 7, 'retry_backoff': 5 },
)
def download_illustration(
    payload: DownloadIllustrationTaskArgs,
    downloader: IDownloader,
) -> None:
    downloader.download(task_id=payload.task_id)
```

FastAPI provide great developer experience when you develop web API but when the application starts grow and need to integrate with other tools/libraries it's difficult to keep the same quality of architecture for whole project.

**Comment by arpanpreneur** [comment]
@mike-optifino Could you please share how your example exactly works? I am trying to accomplish exactly that.

**Reply by mike-optifino** [reply]
it will be big amount of code, not sure that this is right place to share it, but i can describe the concept. 

1. Implement `TaskRegistry` aka `APIRoute` from FastAPI. This class provide ability to use it as decorator and remember all functions that was decorated
2. Add hook to generate upgraded celery task when celery worker running   
3. Use next function to generate task
```
def generate_celery_task(
    task_annotation: TaskAnnotation,
    celery_app: Celery,
    dependency_provider: IDependencyProvider,
) -> BaseTask:
    task_kwargs = task_annotation.get_task_kwargs()
    task_name = task_kwargs.get('name')
    task_bind = task_kwargs.pop('bind')
    assert task_name
    assert callable(task_annotation.dependant.call), 'An task must be a callable'

    @celery_app.task(bind=True, track_started=True, **task_kwargs)
    def wrapper(*args: Any, **kwargs: dict[str, Any]) -> None:
        assert args
        task: BaseTask = args[0]
        with dependency_provider as (dependency_cache, stack):
            resolved_dependencies, errors, _ = solve_dependencies(
                dependant=task_annotation.dependant,
                dependency_cache=dependency_cache,
                stack=stack,
                dependency_provider=dependency_provider,
            )
            if errors:
                raise ValueError(f'Invalid dependencies: {errors}')

            if task_bind and task_argument_name:
                resolved_dependencies[task_argument_name] = task

            if task_annotation.task_arg_cls is not None and task_annotation.task_arg_name is not None:
                task_arg_cls = task_annotation.task_arg_cls
                task_args_dict: dict[str, Any] = args[1]
                task_typed_arg = task_arg_cls(**task_args_dict)
                resolved_dependencies[task_annotation.task_arg_name] = task_typed_arg

            task_annotation.dependant.call(**resolved_dependencies)

    return cast(BaseTask, wrapper)
```
4. I duplicate code of `solve_dependencies` function and get rid of http details

From my POV it's not good solution but it's allow to provide consistent DX in entire project.
@arpanpreneur sorry for the late reply

**Comment by cjw296** [comment]
Did FastAPI grow something like this built-in yet? I can't find anything in the docs :-/ 
I guess maybe https://fastapi.tiangolo.com/advanced/events/#lifespan-events?
